File: server/clients/realdebrid_client.py
"""
Real-Debrid API client.

Relevant endpoints used
-----------------------
POST /torrents/addMagnet          → add a magnet to RD; returns torrent id
POST /torrents/selectFiles/{id}   → select all files for download
GET  /torrents/info/{id}          → poll status; when "downloaded" fetch links[]
POST /unrestrict/link             → convert an RD link to a direct CDN URL
GET  /torrents/instantAvailability/{hash}  → check cache before adding

Real-Debrid base URL: https://api.real-debrid.com/rest/1.0
"""
from __future__ import annotations

import asyncio
import logging
from typing import Optional

import httpx

logger = logging.getLogger(__name__)

_MAX_RETRIES = 2
_RETRY_BACKOFF = 2

RD_BASE = "https://api.real-debrid.com/rest/1.0"

# Torrent statuses considered "done" on RD's side
_DONE_STATUSES = {"downloaded", "magnet_done"}
_ERROR_STATUSES = {"error", "virus", "dead", "magnet_error"}


class RealDebridError(Exception):
    pass


class RealDebridClient:
    def __init__(self, api_key: str, poll_interval: int = 30):
        self._key = api_key
        self._poll_interval = poll_interval
        self._client = httpx.AsyncClient(
            base_url=RD_BASE,
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=15,
        )

    async def close(self) -> None:
        await self._client.aclose()

    # ------------------------------------------------------------------
    # Cache check
    # ------------------------------------------------------------------

    async def is_cached(self, info_hash: str) -> bool:
        """Return True if the torrent hash is instantly available on RD.

        Retries on transient connection errors so a single network hiccup
        doesn't cause every hash to look uncached.
        """
        last_exc: Optional[Exception] = None

        for attempt in range(_MAX_RETRIES):
            try:
                r = await self._client.get(
                    f"/torrents/instantAvailability/{info_hash.lower()}"
                )
                r.raise_for_status()
                data = r.json()
                entry = data.get(info_hash.lower(), {})
                return bool(entry.get("rd"))
            except (httpx.ConnectError, httpx.ConnectTimeout, httpx.ReadTimeout) as exc:
                last_exc = exc
                delay = _RETRY_BACKOFF * (2 ** attempt)
                logger.info(
                    "RD cache check attempt %d/%d failed (%s) — retrying in %ds",
                    attempt + 1, _MAX_RETRIES, type(exc).__name__, delay,
                )
                await asyncio.sleep(delay)
            except Exception as exc:
                logger.warning("RD cache check failed: %s", exc)
                return False

        logger.warning("RD cache check failed after %d retries: %s", _MAX_RETRIES, last_exc)
        return False

    # ------------------------------------------------------------------
    # Add magnet
    # ------------------------------------------------------------------

    async def add_magnet(self, magnet: str) -> str:
        """Add *magnet* to RD and return the RD torrent ID."""
        r = await self._client.post("/torrents/addMagnet", data={"magnet": magnet})
        if r.status_code not in (200, 201):
            raise RealDebridError(f"addMagnet failed ({r.status_code}): {r.text}")
        data = r.json()
        torrent_id = data.get("id")
        if not torrent_id:
            raise RealDebridError(f"addMagnet returned no id: {data}")
        logger.info("RD torrent added: %s", torrent_id)
        return torrent_id

    # ------------------------------------------------------------------
    # Select files
    # ------------------------------------------------------------------

    async def select_all_files(self, torrent_id: str) -> None:
        """Tell RD to download all files in the torrent."""
        r = await self._client.post(
            f"/torrents/selectFiles/{torrent_id}",
            data={"files": "all"},
        )
        if r.status_code not in (200, 201, 204):
            raise RealDebridError(
                f"selectFiles failed ({r.status_code}): {r.text}"
            )

    # ------------------------------------------------------------------
    # Poll until downloaded
    # ------------------------------------------------------------------

    async def wait_until_downloaded(
        self,
        torrent_id: str,
        on_progress=None,   # optional async callable(percent: int)
    ) -> list[str]:
        """
        Poll the RD torrent info endpoint until the torrent is downloaded.
        Returns the list of RD download links.
        """
        while True:
            info = await self.get_torrent_info(torrent_id)
            status = info.get("status", "")
            progress = info.get("progress", 0)

            logger.debug("RD torrent %s status=%s progress=%s%%", torrent_id, status, progress)

            if on_progress:
                await on_progress(int(progress))

            if status in _DONE_STATUSES:
                links = info.get("links", [])
                if not links:
                    raise RealDebridError("Torrent downloaded but no links returned")
                return links

            if status in _ERROR_STATUSES:
                raise RealDebridError(f"RD torrent failed with status: {status}")

            await asyncio.sleep(self._poll_interval)

    async def get_torrent_info(self, torrent_id: str) -> dict:
        r = await self._client.get(f"/torrents/info/{torrent_id}")
        r.raise_for_status()
        return r.json()

    # ------------------------------------------------------------------
    # Unrestrict link → direct CDN download URL
    # ------------------------------------------------------------------

    async def unrestrict_link(self, link: str) -> tuple[str, Optional[int]]:
        """
        Convert an RD link to a direct CDN download URL.
        Returns (download_url, filesize_bytes).
        """
        r = await self._client.post("/unrestrict/link", data={"link": link})
        if r.status_code not in (200, 201):
            raise RealDebridError(f"unrestrict/link failed ({r.status_code}): {r.text}")
        data = r.json()
        url = data.get("download") or data.get("url")
        size = data.get("filesize")
        if not url:
            raise RealDebridError(f"unrestrict/link returned no URL: {data}")
        return url, size

    async def unrestrict_all(self, links: list[str]) -> list[tuple[str, Optional[int]]]:
        """Unrestrict a list of RD links concurrently."""
        tasks = [self.unrestrict_link(link) for link in links]
        return list(await asyncio.gather(*tasks))

    # ------------------------------------------------------------------
    # High-level: add + select + wait + unrestrict in one call
    # ------------------------------------------------------------------

    async def download_magnet(
        self, magnet: str, on_progress=None
    ) -> tuple[list[tuple[str, Optional[int]]], str]:
        """
        Full pipeline for a non-cached magnet:
          1. Add to RD
          2. Select all files
          3. Poll until downloaded
          4. Unrestrict all links
        Returns (list_of_(cdn_url, file_size_bytes), torrent_id).
        """
        torrent_id = await self.add_magnet(magnet)
        await self.select_all_files(torrent_id)
        rd_links = await self.wait_until_downloaded(torrent_id, on_progress=on_progress)
        unrestricted = await self.unrestrict_all(rd_links)
        return unrestricted, torrent_id

File: app/MediaDownloader/MediaDownloader.Server/Clients/RealDebridClient.cs
using System.Text.Json;

namespace MediaDownloader.Server.Clients;

public class RealDebridError : Exception
{
    public RealDebridError(string message) : base(message) { }
}

/// <summary>
/// Real-Debrid API client. Mirrors server/clients/realdebrid_client.py.
/// </summary>
public class RealDebridClient : BaseApiClient
{
    private const string RdBase = "https://api.real-debrid.com/rest/1.0";
    private static readonly HashSet<string> DoneStatuses = ["downloaded", "magnet_done"];
    private static readonly HashSet<string> ErrorStatuses = ["error", "virus", "dead", "magnet_error"];
    private const int MaxWaitSeconds = 30 * 60;

    private readonly int _pollInterval;

    public RealDebridClient(string apiKey, int pollInterval = 30, ILogger? logger = null)
        : base(timeoutSeconds: 15, maxRetries: 3, backoffBase: 2.0,
              headers: new() { ["Authorization"] = $"Bearer {apiKey}" }, logger: logger)
    {
        _pollInterval = pollInterval;
    }

    public async Task<bool> IsCachedAsync(string infoHash, CancellationToken ct = default)
    {
        try
        {
            var data = await GetJsonAsync($"{RdBase}/torrents/instantAvailability/{infoHash.ToLower()}", ct);
            if (data.TryGetProperty(infoHash.ToLower(), out var entry) &&
                entry.TryGetProperty("rd", out var rd))
                return rd.GetArrayLength() > 0;
            return false;
        }
        catch { return false; }
    }

    public async Task<string> AddMagnetAsync(string magnet, CancellationToken ct = default)
    {
        var content = new FormUrlEncodedContent(new[] { new KeyValuePair<string, string>("magnet", magnet) });
        var response = await PostAsync($"{RdBase}/torrents/addMagnet", content, ct);
        var json = await response.Content.ReadAsStringAsync(ct);
        var data = JsonSerializer.Deserialize<JsonElement>(json);
        var torrentId = data.TryGetProperty("id", out var id) ? id.GetString() : null;
        if (torrentId == null)
            throw new RealDebridError($"addMagnet returned no id: {json}");
        return torrentId;
    }

    public async Task SelectAllFilesAsync(string torrentId, CancellationToken ct = default)
    {
        var content = new FormUrlEncodedContent(new[] { new KeyValuePair<string, string>("files", "all") });
        await PostAsync($"{RdBase}/torrents/selectFiles/{torrentId}", content, ct);
    }

    public async Task<List<string>> WaitUntilDownloadedAsync(
        string torrentId, Func<int, Task>? onProgress = null,
        int maxWait = MaxWaitSeconds, CancellationToken ct = default)
    {
        var deadline = Environment.TickCount64 + maxWait * 1000L;

        while (true)
        {
            if (Environment.TickCount64 > deadline)
                throw new RealDebridError($"RD download timed out after {maxWait}s for torrent {torrentId}");

            var info = await GetTorrentInfoAsync(torrentId, ct);
            var status = info.TryGetProperty("status", out var s) ? s.GetString() ?? "" : "";
            var progress = info.TryGetProperty("progress", out var p) ? p.GetInt32() : 0;

            if (onProgress != null)
                await onProgress(progress);

            if (DoneStatuses.Contains(status))
            {
                var links = info.TryGetProperty("links", out var l)
                    ? l.EnumerateArray().Select(x => x.GetString() ?? "").Where(x => x.Length > 0).ToList()
                    : new List<string>();
                if (links.Count == 0)
                    throw new RealDebridError("Torrent downloaded but no links returned");
                return links;
            }

            if (ErrorStatuses.Contains(status))
                throw new RealDebridError($"RD torrent failed with status: {status}");

            await Task.Delay(TimeSpan.FromSeconds(_pollInterval), ct);
        }
    }

    public async Task<JsonElement> GetTorrentInfoAsync(string torrentId, CancellationToken ct = default)
        => await GetJsonAsync($"{RdBase}/torrents/info/{torrentId}", ct);

    public async Task<(string url, long? size)> UnrestrictLinkAsync(string link, CancellationToken ct = default)
    {
        var content = new FormUrlEncodedContent(new[] { new KeyValuePair<string, string>("link", link) });
        var data = await PostJsonAsync($"{RdBase}/unrestrict/link", content, ct);
        var url = data.TryGetProperty("download", out var d) ? d.GetString()
                : data.TryGetProperty("url", out var u) ? u.GetString() : null;
        var size = data.TryGetProperty("filesize", out var fs) ? fs.GetInt64() : (long?)null;
        if (url == null)
            throw new RealDebridError($"unrestrict/link returned no URL");
        return (url, size);
    }

    public async Task<List<(string url, long? size)>> UnrestrictAllAsync(List<string> links, CancellationToken ct = default)
    {
        var tasks = links.Select(link => UnrestrictLinkAsync(link, ct));
        return (await Task.WhenAll(tasks)).ToList();
    }

    public async Task<(List<(string url, long? size)> files, string torrentId)> DownloadMagnetAsync(
        string magnet, Func<int, Task>? onProgress = null, CancellationToken ct = default)
    {
        var torrentId = await AddMagnetAsync(magnet, ct);
        await SelectAllFilesAsync(torrentId, ct);
        var rdLinks = await WaitUntilDownloadedAsync(torrentId, onProgress, ct: ct);
        var unrestricted = await UnrestrictAllAsync(rdLinks, ct);
        return (unrestricted, torrentId);
    }
}
